Failing getOptimalPosition tests on master

https://travis-ci.org/ckeditor/ckeditor5/builds/181701427

A follow-up of https://github.com/ckeditor/ckeditor5-ui-default/issues/137.

Those tests still fail for me in this case:

```
gulp test --files=core,enter,list,typing,undo,utils
```

Failed ones:

```
✖ 14 tests failed

FAILED TESTS:
  getOptimalPosition
    for single position
      ✖ should return coordinates
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 80, top: 491, name: 'left' } to deeply equal { top: 100, left: 80, name: 'left' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:94:5)

    for multiple positions
      ✖ should return coordinates
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 80, top: 491, name: 'left' } to deeply equal { top: 100, left: 80, name: 'left' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:145:5)

      ✖ should return coordinates (position preference order)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 110, top: 491, name: 'right' } to deeply equal { top: 100, left: 110, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:156:5)

    with a limiter
      ✖ should return coordinates (#1)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: -20, top: 491, name: 'left' } to deeply equal { top: 100, left: -20, name: 'left' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:171:5)

      ✖ should return coordinates (#2)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: -20, top: 491, name: 'left' } to deeply equal { top: 100, left: -20, name: 'left' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:182:5)

    with fitInViewport on
      ✖ should return coordinates (#1)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 10, top: 491, name: 'right' } to deeply equal { top: 100, left: 10, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:197:5)

      ✖ should return coordinates (#2)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 10, top: 491, name: 'right' } to deeply equal { top: 100, left: 10, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:209:5)

      ✖ should return coordinates (#3)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 0, top: 501, name: 'bottom' } to deeply equal { top: 110, left: 0, name: 'bottom' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:221:5)

    with limiter and fitInViewport on
      ✖ should return coordinates (#1)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 10, top: 491, name: 'right' } to deeply equal { top: 100, left: 10, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:237:5)

      ✖ should return coordinates (#2)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 10, top: 491, name: 'right' } to deeply equal { top: 100, left: 10, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:249:5)

      ✖ should return coordinates (#3)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 0, top: 501, name: 'bottom' } to deeply equal { top: 110, left: 0, name: 'bottom' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:261:5)

      ✖ should return coordinates (#4)
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 10, top: 491, name: 'right' } to deeply equal { top: 100, left: 10, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:273:5)

      ✖ should return the very first coordinates if no fitting position with a positive intersection has been found
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { Object (left, top, ...) } to deeply equal { Object (left, top, ...) }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:285:5)

      ✖ should return the very first coordinates if limiter does not fit into the viewport
        Chrome 55.0.2883 (Mac OS X 10.12.1)
      AssertionError: expected { left: 10, top: 491, name: 'right' } to deeply equal { top: 100, left: 10, name: 'right' }
          at Assertion.assertEqual (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:774:19)
          at Assertion.ctx.(anonymous function) [as equal] (/www/ckeditor5/ckeditor5/node_modules/chai/chai.js:4192:25)
          at assertPosition (tests/utils/dom/position.js:326:28)
          at Context.it (tests/utils/dom/position.js:310:5)

```

--- a/src/dom/getpositionedancestor.js
+++ b/src/dom/getpositionedancestor.js
@@ -3,11 +3,13 @@
  * For licensing, see LICENSE.md.
  */
 
-/* globals window */
-
 /**
  * @module utils/dom/getpositionedancestor
  */
+
+import global from './global.js';
+
+const window = global.window;
 
 /**
  * For a given element, returns the nearest ancestor element which CSS position is not "static".

--- a/src/dom/getpositionedancestor.js
+++ b/src/dom/getpositionedancestor.js
@@ -9,8 +9,6 @@
 
 import global from './global.js';
 
-const window = global.window;
-
 /**
  * For a given element, returns the nearest ancestor element which CSS position is not "static".
  *
@@ -19,7 +17,7 @@ const window = global.window;
  */
 export default function getPositionedAncestor( element ) {
 	while ( element && element.tagName.toLowerCase() != 'html' ) {
-		if ( window.getComputedStyle( element ).position != 'static' ) {
+		if ( global.window.getComputedStyle( element ).position != 'static' ) {
 			return element;
 		}
 

--- a/src/dom/global.js
+++ b/src/dom/global.js
@@ -16,11 +16,11 @@
  *
  *		import global from 'ckeditor5/utils/dom/global.js';
  *
- *		const window = global.window;
- *
  *		// This stub will work for any code using global module.
  *		testUtils.sinon.stub( global, 'window', {
  *			innerWidth: 10000
  *		} );
+ *
+ *		console.log( global.window.innerWidth );
  */
 export default { window, document };

--- a/src/dom/global.js
+++ b/src/dom/global.js
@@ -12,13 +12,15 @@
 /**
  * A helper (module) giving an access to the global DOM objects such as `window` and
  * `document`. Accessing these objects using this helper allows easy and bulletproof
- * testing, i.e. stubbing native properties
+ * testing, i.e. stubbing native properties:
  *
  *		import global from 'ckeditor5/utils/dom/global.js';
  *
  *		const window = global.window;
  *
- * 		// This stub will work for any code using global module in tests.
- *		testUtils.sinon.stub( window, 'innerWidth', 10000 );
+ *		// This stub will work for any code using global module.
+ *		testUtils.sinon.stub( global, 'window', {
+ *			innerWidth: 10000
+ *		} );
  */
-export default { document, window };
+export default { window, document };

--- a/src/dom/global.js
+++ b/src/dom/global.js
@@ -1,0 +1,24 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/* globals window, document */
+
+/**
+ * @module utils/dom/global
+ */
+
+/**
+ * A helper (module) giving an access to the global DOM objects such as `window` and
+ * `document`. Accessing these objects using this helper allows easy and bulletproof
+ * testing, i.e. stubbing native properties
+ *
+ *		import global from 'ckeditor5/utils/dom/global.js';
+ *
+ *		const window = global.window;
+ *
+ * 		// This stub will work for any code using global module in tests.
+ *		testUtils.sinon.stub( window, 'innerWidth', 10000 );
+ */
+export default { document, window };

--- a/src/dom/position.js
+++ b/src/dom/position.js
@@ -3,14 +3,15 @@
  * For licensing, see LICENSE.md.
  */
 
-/* globals window */
-
 /**
  * @module utils/dom/position
  */
 
 import Rect from './rect.js';
+import global from './global.js';
 import getPositionedAncestor from './getpositionedancestor.js';
+
+const window = global.window;
 
 /**
  * Calculates the `position: absolute` coordinates of a given element so it can be positioned with respect to the

--- a/src/dom/position.js
+++ b/src/dom/position.js
@@ -7,11 +7,9 @@
  * @module utils/dom/position
  */
 
-import Rect from './rect.js';
 import global from './global.js';
+import Rect from './rect.js';
 import getPositionedAncestor from './getpositionedancestor.js';
-
-const window = global.window;
 
 /**
  * Calculates the `position: absolute` coordinates of a given element so it can be positioned with respect to the
@@ -216,9 +214,11 @@ function getBestPosition( positions, targetRect, elementRect, limiterRect, viewp
 // @param {utils/dom/rect~Rect} rect A rect to be converted.
 // @returns {Object} Object containing `left` and `top` properties, in absolute coordinates.
 function getAbsoluteRectCoordinates( { left, top } ) {
+	const { scrollX, scrollY } = global.window;
+
 	return {
-		left: left + window.scrollX,
-		top: top + window.scrollY,
+		left: left + scrollX,
+		top: top + scrollY,
 	};
 }
 

--- a/src/dom/rect.js
+++ b/src/dom/rect.js
@@ -11,7 +11,6 @@ import global from './global.js';
 import isRange from './isrange.js';
 import isElement from '../../utils/lib/lodash/isElement.js';
 
-const window = global.window;
 const rectProperties = [ 'top', 'right', 'bottom', 'left', 'width', 'height' ];
 
 /**
@@ -186,7 +185,7 @@ export default class Rect {
 	 * @returns {module:utils/dom/rect~Rect} A viewport rect.
 	 */
 	static getViewportRect() {
-		const { innerWidth, innerHeight } = window;
+		const { innerWidth, innerHeight } = global.window;
 
 		return new Rect( {
 			top: 0,

--- a/src/dom/rect.js
+++ b/src/dom/rect.js
@@ -3,15 +3,15 @@
  * For licensing, see LICENSE.md.
  */
 
-/* global window */
-
 /**
  * @module utils/dom/rect
  */
 
+import global from './global.js';
 import isRange from './isrange.js';
 import isElement from '../../utils/lib/lodash/isElement.js';
 
+const window = global.window;
 const rectProperties = [ 'top', 'right', 'bottom', 'left', 'width', 'height' ];
 
 /**

--- a/tests/dom/global.js
+++ b/tests/dom/global.js
@@ -1,0 +1,24 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/* global window, document */
+
+import global from 'ckeditor5/utils/dom/global.js';
+
+describe( 'global', () => {
+	describe( 'global', () => {
+		describe( 'window', () => {
+			it( 'equals native DOM window', () => {
+				expect( global.window ).to.equal( window );
+			} );
+		} );
+
+		describe( 'document', () => {
+			it( 'equals native DOM document', () => {
+				expect( global.document ).to.equal( document );
+			} );
+		} );
+	} );
+} );

--- a/tests/dom/global.js
+++ b/tests/dom/global.js
@@ -6,6 +6,9 @@
 /* global window, document */
 
 import global from 'ckeditor5/utils/dom/global.js';
+import testUtils from 'tests/core/_utils/utils.js';
+
+testUtils.createSinonSandbox();
 
 describe( 'global', () => {
 	describe( 'global', () => {
@@ -13,11 +16,31 @@ describe( 'global', () => {
 			it( 'equals native DOM window', () => {
 				expect( global.window ).to.equal( window );
 			} );
+
+			it( 'stubs', () => {
+				testUtils.sinon.stub( global, 'window', {
+					scrollX: 100
+				} );
+
+				expect( global.window ).to.deep.equal( {
+					scrollX: 100
+				} );
+			} );
 		} );
 
 		describe( 'document', () => {
 			it( 'equals native DOM document', () => {
 				expect( global.document ).to.equal( document );
+			} );
+
+			it( 'stubs', () => {
+				testUtils.sinon.stub( global, 'document', {
+					foo: 'abc'
+				} );
+
+				expect( global.document ).to.deep.equal( {
+					foo: 'abc'
+				} );
 			} );
 		} );
 	} );

--- a/tests/dom/position.js
+++ b/tests/dom/position.js
@@ -3,33 +3,25 @@
  * For licensing, see LICENSE.md.
  */
 
-/* global document, window */
+/* global document */
 
+import global from 'ckeditor5/utils/dom/global.js';
 import { getOptimalPosition } from 'ckeditor5/utils/dom/position.js';
-import Rect from 'ckeditor5/utils/dom/rect.js';
 import testUtils from 'tests/core/_utils/utils.js';
 
 testUtils.createSinonSandbox();
 
-let element, target, limiter, revertWindowScroll;
+const window = global.window;
+let element, target, limiter;
 
 describe( 'getOptimalPosition', () => {
 	beforeEach( () => {
 		// Give us a lot of space.
-		testUtils.sinon.stub( Rect, 'getViewportRect' ).returns( new Rect( {
-			top: 0,
-			right: 10000,
-			bottom: 10000,
-			left: 0,
-			width: 10000,
-			height: 10000
-		} ) );
-	} );
+		testUtils.sinon.stub( window, 'innerWidth', 10000 );
+		testUtils.sinon.stub( window, 'innerHeight', 10000 );
 
-	afterEach( () => {
-		if ( revertWindowScroll ) {
-			revertWindowScroll();
-		}
+		testUtils.sinon.stub( window, 'scrollX', 0 );
+		testUtils.sinon.stub( window, 'scrollY', 0 );
 	} );
 
 	describe( 'for single position', () => {
@@ -44,7 +36,8 @@ describe( 'getOptimalPosition', () => {
 		} );
 
 		it( 'should return coordinates (window scroll)', () => {
-			stubWindowScroll( 100, 100 );
+			testUtils.sinon.stub( window, 'scrollX', 100 );
+			testUtils.sinon.stub( window, 'scrollY', 100 );
 
 			assertPosition( { element, target, positions: [ attachLeft ] }, {
 				top: 200,
@@ -55,7 +48,9 @@ describe( 'getOptimalPosition', () => {
 
 		it( 'should return coordinates (positioned element parent)', () => {
 			const positionedParent = document.createElement( 'div' );
-			stubWindowScroll( 1000, 1000 );
+
+			testUtils.sinon.stub( window, 'scrollX', 1000 );
+			testUtils.sinon.stub( window, 'scrollY', 1000 );
 
 			Object.assign( positionedParent.style, {
 				position: 'absolute',
@@ -312,20 +307,6 @@ const attachTop = ( targetRect, elementRect ) => ( {
 	left: targetRect.left - ( elementRect.width - targetRect.width ),
 	name: 'bottom'
 } );
-
-function stubWindowScroll( x, y ) {
-	const { scrollX: savedX, scrollY: savedY } = window;
-
-	window.scrollX = x;
-	window.scrollY = y;
-
-	revertWindowScroll = () => {
-		window.scrollX = savedX;
-		window.scrollY = savedY;
-
-		revertWindowScroll = null;
-	};
-}
 
 function stubElementRect( element, rect ) {
 	if ( element.getBoundingClientRect.restore ) {

--- a/tests/dom/position.js
+++ b/tests/dom/position.js
@@ -3,7 +3,7 @@
  * For licensing, see LICENSE.md.
  */
 
-/* global document */
+/* global document, window */
 
 import global from 'ckeditor5/utils/dom/global.js';
 import { getOptimalPosition } from 'ckeditor5/utils/dom/position.js';
@@ -11,17 +11,18 @@ import testUtils from 'tests/core/_utils/utils.js';
 
 testUtils.createSinonSandbox();
 
-const window = global.window;
-let element, target, limiter;
+let element, target, limiter, windowStub;
 
 describe( 'getOptimalPosition', () => {
 	beforeEach( () => {
-		// Give us a lot of space.
-		testUtils.sinon.stub( window, 'innerWidth', 10000 );
-		testUtils.sinon.stub( window, 'innerHeight', 10000 );
+		windowStub = {
+			innerWidth: 10000,
+			innerHeight: 10000,
+			scrollX: 0,
+			scrollY: 0
+		};
 
-		testUtils.sinon.stub( window, 'scrollX', 0 );
-		testUtils.sinon.stub( window, 'scrollY', 0 );
+		testUtils.sinon.stub( global, 'window', windowStub );
 	} );
 
 	describe( 'for single position', () => {
@@ -36,8 +37,12 @@ describe( 'getOptimalPosition', () => {
 		} );
 
 		it( 'should return coordinates (window scroll)', () => {
-			testUtils.sinon.stub( window, 'scrollX', 100 );
-			testUtils.sinon.stub( window, 'scrollY', 100 );
+			Object.assign( windowStub, {
+				innerWidth: 10000,
+				innerHeight: 10000,
+				scrollX: 100,
+				scrollY: 100,
+			} );
 
 			assertPosition( { element, target, positions: [ attachLeft ] }, {
 				top: 200,
@@ -49,8 +54,15 @@ describe( 'getOptimalPosition', () => {
 		it( 'should return coordinates (positioned element parent)', () => {
 			const positionedParent = document.createElement( 'div' );
 
-			testUtils.sinon.stub( window, 'scrollX', 1000 );
-			testUtils.sinon.stub( window, 'scrollY', 1000 );
+			Object.assign( windowStub, {
+				innerWidth: 10000,
+				innerHeight: 10000,
+				scrollX: 1000,
+				scrollY: 1000,
+				getComputedStyle: ( el ) => {
+					return window.getComputedStyle( el );
+				}
+			} );
 
 			Object.assign( positionedParent.style, {
 				position: 'absolute',

--- a/tests/dom/rect.js
+++ b/tests/dom/rect.js
@@ -3,10 +3,13 @@
  * For licensing, see LICENSE.md.
  */
 
-/* global document, window */
+/* global document */
 
+import global from 'ckeditor5/utils/dom/global.js';
 import Rect from 'ckeditor5/utils/dom/rect.js';
 import testUtils from 'tests/core/_utils/utils.js';
+
+const window = global.window;
 
 testUtils.createSinonSandbox();
 
@@ -325,10 +328,11 @@ describe( 'Rect', () => {
 		} );
 
 		it( 'should return the viewport\'s rect', () => {
-			window.scrollX = 100;
-			window.scrollY = 200;
-			window.innerWidth = 1000;
-			window.innerHeight = 500;
+			testUtils.sinon.stub( window, 'innerWidth', 1000 );
+			testUtils.sinon.stub( window, 'innerHeight', 500 );
+
+			testUtils.sinon.stub( window, 'scrollX', 100 );
+			testUtils.sinon.stub( window, 'scrollY', 200 );
 
 			assertRect( Rect.getViewportRect(), {
 				top: 0,

--- a/tests/dom/rect.js
+++ b/tests/dom/rect.js
@@ -9,8 +9,6 @@ import global from 'ckeditor5/utils/dom/global.js';
 import Rect from 'ckeditor5/utils/dom/rect.js';
 import testUtils from 'tests/core/_utils/utils.js';
 
-const window = global.window;
-
 testUtils.createSinonSandbox();
 
 describe( 'Rect', () => {
@@ -328,11 +326,12 @@ describe( 'Rect', () => {
 		} );
 
 		it( 'should return the viewport\'s rect', () => {
-			testUtils.sinon.stub( window, 'innerWidth', 1000 );
-			testUtils.sinon.stub( window, 'innerHeight', 500 );
-
-			testUtils.sinon.stub( window, 'scrollX', 100 );
-			testUtils.sinon.stub( window, 'scrollY', 200 );
+			testUtils.sinon.stub( global, 'window', {
+				innerWidth: 1000,
+				innerHeight: 500,
+				scrollX: 100,
+				scrollY: 200
+			} );
 
 			assertRect( Rect.getViewportRect(), {
 				top: 0,

